File: long_term_memory.py
import os
from pathlib import Path
from datetime import datetime
from typing import List, Dict

class LongTermMemory:
    def __init__(self, log_dir="LongTermMemoryLog"):
        self.log_dir = Path(log_dir)
        self.log_dir.mkdir(exist_ok=True)
        self.entries = []
        self.load_all_logs()

    def load_all_logs(self):
        """Load all long-term memory entries from log files."""
        self.entries = []
        for file_path in self.log_dir.glob("*.txt"):
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read().strip()
                    if content:
                        self.entries.append({
                            "filename": file_path.name,
                            "content": content,
                            "date": self._extract_date(content)
                        })
            except Exception as e:
                print(f"Error loading {file_path}: {e}")

    def _extract_date(self, content: str) -> str:
        """Extract date from content if present."""
        for line in content.split('\n'):
            if line.startswith("Date:"):
                return line.split("Date:")[1].strip()
        return datetime.now().strftime("%Y-%m-%d")

    def save_entry(self, entry: str, category: str = "General"):
        """Save a new long-term memory entry."""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        filename = f"{timestamp}_{category}_Memory--FENR1R.txt"
        filepath = self.log_dir / filename

        content = f"Long-Term Memory Entry: {category}\nDate: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n\n{entry}\n\nCitation: Auto-saved from chat interaction."

        try:
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(content)
            self.entries.append({
                "filename": filename,
                "content": content,
                "date": datetime.now().strftime("%Y-%m-%d")
            })
            print(f"Saved long-term memory: {filename}")
        except Exception as e:
            print(f"Error saving long-term memory: {e}")

    def get_relevant_memories(self, query: str, limit: int = 5) -> List[str]:
        """Retrieve relevant long-term memories based on query."""
        # Simple keyword matching; could be improved with embeddings
        relevant = []
        query_lower = query.lower()
        for entry in self.entries:
            if any(word in entry["content"].lower() for word in query_lower.split()):
                relevant.append(entry["content"])
                if len(relevant) >= limit:
                    break
        return relevant

    def get_all_summaries(self) -> str:
        """Get a summary of all long-term memories."""
        if not self.entries:
            return "No long-term memories available."
        summaries = [f"Memory {i+1} ({entry['date']}): {entry['content'][:200]}..." for i, entry in enumerate(self.entries)]
        return "\n".join(summaries)

File: LongTermMemory.cs
using System.Text.RegularExpressions;

namespace FENR1R;

public sealed class LongTermMemory
{
    private readonly string _logDir;
    private readonly List<MemoryEntry> _entries = new();

    public LongTermMemory(string logDir = "LongTermMemoryLog")
    {
        _logDir = logDir;
        Directory.CreateDirectory(_logDir);
        LoadAllLogs();
    }

    public void LoadAllLogs()
    {
        _entries.Clear();
        foreach (var filePath in Directory.EnumerateFiles(_logDir, "*.txt"))
        {
            try
            {
                var content = File.ReadAllText(filePath);
                if (!string.IsNullOrWhiteSpace(content))
                {
                    _entries.Add(new MemoryEntry
                    {
                        Filename = Path.GetFileName(filePath),
                        Content = content,
                        Date = ExtractDate(content)
                    });
                }
            }
            catch
            {
                // ignore invalid log files
            }
        }
    }

    public void SaveEntry(string entry, string category = "General")
    {
        var timestamp = DateTime.UtcNow.ToString("yyyyMMdd_HHmmss");
        var filename = $"{timestamp}_{category}_Memory--FENR1R.txt";
        var filepath = Path.Combine(_logDir, filename);
        var content = $"Long-Term Memory Entry: {category}\nDate: {DateTime.UtcNow:yyyy-MM-dd HH:mm:ss}\n\n{entry}\n\nCitation: Auto-saved from chat interaction.";

        File.WriteAllText(filepath, content);
        _entries.Add(new MemoryEntry
        {
            Filename = filename,
            Content = content,
            Date = DateTime.UtcNow.ToString("yyyy-MM-dd")
        });
    }

    public IReadOnlyList<string> GetRelevantMemories(string query, int limit = 5)
    {
        if (string.IsNullOrWhiteSpace(query))
        {
            return Array.Empty<string>();
        }

        var queryTerms = query.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(term => term.ToLowerInvariant())
            .ToHashSet();

        var relevant = new List<string>();
        foreach (var entry in _entries)
        {
            var lowerContent = entry.Content.ToLowerInvariant();
            if (queryTerms.Any(term => lowerContent.Contains(term)))
            {
                relevant.Add(entry.Content);
            }

            if (relevant.Count >= limit)
            {
                break;
            }
        }

        return relevant;
    }

    public string GetAllSummaries()
    {
        if (_entries.Count == 0)
        {
            return "No long-term memories available.";
        }

        return string.Join("\n", _entries.Select((entry, index) =>
            $"Memory {index + 1} ({entry.Date}): {entry.Content[..Math.Min(200, entry.Content.Length)]}..."));
    }

    private static string ExtractDate(string content)
    {
        var match = Regex.Match(content, @"^Date:\s*(.+)$", RegexOptions.Multiline);
        return match.Success ? match.Groups[1].Value.Trim() : DateTime.UtcNow.ToString("yyyy-MM-dd");
    }

    private sealed class MemoryEntry
    {
        public string Filename { get; init; } = string.Empty;
        public string Content { get; init; } = string.Empty;
        public string Date { get; init; } = string.Empty;
    }
}
